src/listener_manager.c

```diff
@@ -0,0 +1,138 @@
+// SPDX-License-Identifier: BSD-3-Clause
+/**
+ * @file listener_manager.c
+ *
+ * @brief Map of MPTCP local address ID to listener.
+ *
+ * Copyright (c) 2022, Intel Corporation
+ */
+
+#ifdef HAVE_CONFIG_H
+# include <mptcpd/private/config.h>
+#endif
+
+#include <assert.h>
+#include <sys/types.h>
+#include <sys/socket.h>
+#include <netinet/in.h>
+#include <unistd.h>
+#include <errno.h>
+#include <string.h>
+
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wpedantic"
+#include <ell/hashmap.h>
+#include <ell/util.h>
+#include <ell/log.h>
+#pragma GCC diagnostic pop
+
+#include "listener_manager.h"
+
+
+// ----------------------------------------------------------------------
+
+struct l_hashmap *mptcpd_lm_create(void)
+{
+        // Map of MPTCP local address ID to MPTCP listener file
+        // descriptor.
+        return l_hashmap_new();
+}
+
+static void close_listener(void *value)
+{
+        /*
+          We don't need to worry about overflow due to casting from
+          unsigned int since the file descriptor value stored in the
+          map will never be larger than INT_MAX by design.
+        */
+        int const fd = L_PTR_TO_UINT(value);
+
+        (void) close(fd);
+}
+
+
+void mptcpd_lm_destroy(struct l_hashmap *lm)
+{
+        if (lm == NULL)
+                return;
+
+        l_hashmap_destroy(lm, close_listener);
+}
+
+bool mptcpd_lm_listen(struct l_hashmap *lm,
+                      mptcpd_aid_t id,
```

Comment:
I think the id is extra information here - path managers may choose to assign different address IDs for the same address on different connections but still want to share the listener socket.

Hopefully there's enough information in the `struct sockaddr` alone to track the listener. I see below that `id` is used as the hash key. Could use a custom `l_hashmap_hash_func_t` to hash on sockaddr contents? Unfortunately `hash_superfast()` is not exported from ELL.

We also talked about allowing sharing of the listen sockets, using some kind of reference counting. Then the hash object would have to be a dynamically-allocated struct with the ref count and the socket FD. This struct would also be useful for adding other data, such as tracking `l_io` handles for clearing accept queues on the listeners.

Reply:
> I think the id is extra information here - path managers may choose to assign different address IDs for the same address on different connections but still want to share the listener socket.

Agreed.

> Hopefully there's enough information in the `struct sockaddr` alone to track the listener. I see below that `id` is used as the hash key. Could use a custom `l_hashmap_hash_func_t` to hash on sockaddr contents? Unfortunately `hash_superfast()` is not exported from ELL.

The mptcpd ID manager, for example, uses the IP address as the key.  I assume we'd want to include the port for the mptcpd listener manager map case.  Besides ELL's `hash_superfast()`, I briefly looked into [MurmurHash](https://github.com/aappleby/smhasher/wiki/MurmurHash) for mptcpd's ID manager, but never implemented it since there wasn't a pressing need for it.  According to the author, [MurmurHash performs better](https://github.com/aappleby/smhasher/wiki/MurmurHash3) than [SuperFastHash](http://www.azillionmonkeys.com/qed/hash.html) (same as ELL's `hash_superfast()`).  The licensing for the MurmurHash source code appears to be more liberal, too, i.e. public domain vs LGPL 2.1 for SuperFastHash.


> We also talked about allowing sharing of the listen sockets, using some kind of reference counting. Then the hash object would have to be a dynamically-allocated struct with the ref count and the socket FD. This struct would also be useful for adding other data, such as tracking `l_io` handles for clearing accept queues on the listeners.

That works for me.  Thanks Mat!

Reply:
#242 adds a MurmurHash3 implementation to mptcpd.  Assuming the changes are approved, we could use the new MurmurHash implementation here as well.

Reply:
> Hopefully there's enough information in the struct sockaddr alone to track the listener. I see below that id is used as the hash key. Could use a custom l_hashmap_hash_func_t to hash on sockaddr contents? 

One problem with using a `sockaddr`  as the hash key is that it isn't available when doing a `REMOVE_ADDR` since a MPTCP connection token and address ID are what's needed for that MPTCP generic netlink command.  The corresponding function in `src/netlink_pm_upstream.c` has the following parameters:
```c
static int upstream_remove(struct mptcpd_pm *pm,
                           mptcpd_aid_t id,
                           mptcpd_token_t token)
```

Reply:
If listener sockets are shared and the mptcpd core code handles listener creation, there are several things to consider:
- Address IDs are scoped per-connection. Different connections could advertise the same address and port as different IDs. 
- `mptcpd_lm_listen()` needs to be able to do a lookup to see if there's an existing listener (and add to a refcount if there is).
- `mptcpd_lm_close()` needs to also lookup the socket to decrease the refcount, and close the listener if the refcount goes to 0.

Since `mptcp_lm_listen()` needs to lookup based on `sockaddr`, that seems to be the only hash key that works for both listen/add_addr and close/remove_addr.

You could add a `sockaddr` param to `mptcpd_pm_remove_addr()` to use for the listen socket lookup. I guess that would mean also adding it to `upstream_remove()` because the mptcp_org kernel doesn't need listener sockets.

---

Have you considered exposing the listener manager to the plugins? That way they could create a persistent listener socket (or pool of them) independent of any connection activity, and advertise using such listeners without having to create or destroy them repeatedly. I think this also provides more motivation for making the listener manager API work in terms of things that matter for the listener sockets (like the `sockaddr`s) and not things that are defined per-connection (like address ids).

I think this would allow for automatic listener handling for simple plugins, or more optimized behavior (persistent listener management) for plugins that need it.

Reply:
All of these suggestions and concerns should now be addressed.